api/admin/controller.py

```diff
@@ -2410,26 +2425,93 @@ def _sitewide_settings_controller(self, configuration_object):
+        collection if it has a protocol.
+        """
+        provider_apis = list(self.PROVIDER_APIS)
+        provider_apis.append(OPDSImportMonitor)
+
+        self_test_results = None
+        protocol = protocolClass
+
+        if not collection or not collection.protocol:
+            return None
+
+        if collection.protocol == OPDSImportMonitor.PROTOCOL:
+            protocol = OPDSImportMonitor
+
+        if protocol in provider_apis and issubclass(protocol, HasSelfTests):
+            if (collection.protocol == OPDSImportMonitor.PROTOCOL):
+                self_test_results = protocol.prior_test_results(self._db, protocol, self._db, collection, OPDSImporter)
+            else:
+                self_test_results = protocol.prior_test_results(self._db, protocol, self._db, collection)
+
+        return self_test_results
+
+    def collection_self_tests(self, identifier):
+        protocols = self._get_protocols(self.PROVIDER_APIS)
+
+        if not identifier:
+            return Response("No Identifier", 200)
+
+        if flask.request.method == 'GET':
+            collection = dict()
+            protocolClass = None
+            for col in self._db.query(Collection).filter(Collection.id==int(identifier)):
+                collection = dict(
+                    id=col.id,
+                    name=col.name,
+                    protocol=col.protocol,
+                    parent_id=col.parent_id,
+                    settings=dict(external_account_id=col.external_account_id),
+                )
+
+                if col.protocol in [p.get("name") for p in protocols]:
+                    [protocolClass] = [p for p in self.PROVIDER_APIS if p.NAME == col.protocol]
+
+                collection["self_test_results"] = self._get_prior_test_results(col, protocolClass)
+            return dict(collection=collection)
+
+        if flask.request.method == "POST":
+            collection = dict()
+            collectionProtocol = None
+            protocolClass = None
+            for col in self._db.query(Collection).filter(Collection.id==int(identifier)):
+                collection = col
+                collectionProtocol = col.protocol
+
+                if collectionProtocol in [p.get("name") for p in protocols]:
+                    [protocolClassFound] = [p for p in self.PROVIDER_APIS if p.NAME == col.protocol]
+                    protocolClass = protocolClassFound
+
+            if protocolClass:
+                if (collectionProtocol == OPDSImportMonitor.PROTOCOL):
+                    protocolClass = OPDSImportMonitor
+                    value, results = protocolClass.run_self_tests(self._db, protocolClass, self._db, collection, OPDSImporter)
+                else:
+                    value, results = protocolClass.run_self_tests(self._db, protocolClass, self._db, collection)
+
+                if (value):
+                    return Response("Successfully ran new self tests", 200)
```

Comment:
We usually wrap messages like this in `_(..)` in case we ever decide to translate them.

Reply:
Doesn't need to be done now, but if we wanted to optimize this more in the future I think you could actually return the new test results here so the client only has to make one request.

Reply:
I'm guessing only the string is wrapped and not the whole `Response`?

Reply:
right